src/event_loops_bridge.rs

```diff
@@ -117,3 +119,38 @@ fn pop_scheduled_outside_dlg() -> Option<Box<dyn FnOnce()>> {
         queue.pop_front()
     })
 }
+
+pub async fn wake_from<T>(setup_dlg: impl FnOnce(oneshot::Sender<T>)) -> Result<T, oneshot::error::RecvError> {
+    let (tx, rx) = oneshot::channel();
+    setup_dlg(tx);
+    rx.await
+}
+
+pub async fn outside<T: 'static>(fut: impl Future<Output = T> + 'static) -> Result<T, oneshot::error::RecvError> {
+    wake_from(|tx| {
+        glib::MainContext::ref_thread_default().spawn_local(async move {
+            let result = fut.await;
+            tx.send(result).map_err(|_| "Unable to send future result").unwrap();
+        });
+    })
+    .await
+}
+
+pub async fn run_dialog(
+    dialog: &(impl gtk::DialogExt + gtk::GtkWindowExt + gtk::WidgetExt),
```

Comment:
Maybe use `impl IsA<Dialog>` or something like that instead?

Reply:
Doesn't work:
```rust
error[E0599]: the method `set_modal` exists for reference `&impl glib::object::IsA<gtk::Dialog>`, but its trait bounds were not satisfied                                                                           
   --> src/event_loops_bridge.rs:144:12                                                                                                                                                                             
    |                                                                                                                                                                                                               
144 |     dialog.set_modal(true);                                                                                                                                                                                   
    |            ^^^^^^^^^ method cannot be called on `&impl glib::object::IsA<gtk::Dialog>` due to unsatisfied trait bounds                                                                                        
    |                                                                                                                                                                                                               
    = note: the following trait bounds were not satisfied:                                                                                                                                                          
            `impl glib::object::IsA<gtk::Dialog>: IsA<gtk::Window>`                                                                                                                                                 
            which is required by `impl glib::object::IsA<gtk::Dialog>: gtk::GtkWindowExt`                                                                                                                           
            `impl glib::object::IsA<gtk::Dialog>: IsA<Popover>`                                                                                                                                                     
            which is required by `impl glib::object::IsA<gtk::Dialog>: gtk::PopoverExt`                                                                                                                             
            `&impl glib::object::IsA<gtk::Dialog>: IsA<gtk::Window>`                                                                                                                                                
            which is required by `&impl glib::object::IsA<gtk::Dialog>: gtk::GtkWindowExt`                                                                                                                          
            `&impl glib::object::IsA<gtk::Dialog>: IsA<Popover>`                                                                                                                                                    
            which is required by `&impl glib::object::IsA<gtk::Dialog>: gtk::PopoverExt`                                                                                                                            
                                                                                                                                                                                                                    
error[E0599]: the method `show` exists for reference `&impl glib::object::IsA<gtk::Dialog>`, but its trait bounds were not satisfied                                                                                
   --> src/event_loops_bridge.rs:145:12                                                                                                                                                                             
    |                                                                                                                                                                                                               
145 |     dialog.show();                                                                                                                                                                                            
    |            ^^^^ method cannot be called on `&impl glib::object::IsA<gtk::Dialog>` due to unsatisfied trait bounds                                                                                             
    |                                                                                                                                                                                                               
    = note: the following trait bounds were not satisfied:                                                                                                                                                          
            `impl glib::object::IsA<gtk::Dialog>: IsA<gtk::Widget>`                                                                                                                                                 
            which is required by `impl glib::object::IsA<gtk::Dialog>: gtk::WidgetExt`                                                                                                                              
            `&impl glib::object::IsA<gtk::Dialog>: IsA<gtk::Widget>`                                                                                                                                                
            which is required by `&impl glib::object::IsA<gtk::Dialog>: gtk::WidgetExt`                                                                                                                             
                                                                                                                                                                                                                    
error[E0599]: the method `close` exists for reference `&impl glib::object::IsA<gtk::Dialog>`, but its trait bounds were not satisfied                                                                               
   --> src/event_loops_bridge.rs:151:24                                                                                                                                                                             
    |                                                                                                                                                                                                               
151 |                 dialog.close();                                                                                                                                                                               
    |                        ^^^^^                                                                                                                                                                                  
    |                                                                                                                                                                                                               
    = note: the following trait bounds were not satisfied:                                                                                                                                                          
            `impl glib::object::IsA<gtk::Dialog>: IsA<gtk::Window>`                                                                                                                                                 
            which is required by `impl glib::object::IsA<gtk::Dialog>: gtk::GtkWindowExt`                                                                                                                           
            `&impl glib::object::IsA<gtk::Dialog>: IsA<gtk::Window>`                                                                                                                                                
            which is required by `&impl glib::object::IsA<gtk::Dialog>: gtk::GtkWindowExt
```

Reply:
That's really weird. Because if you look at Gtk methods that accept other objects, that [exactly how they do it](https://gtk-rs.org/docs/gtk/trait.ContainerExt.html#tymethod.add).

Reply:
Yes, but look at [the implementation](https://docs.rs/gtk/0.9.2/src/gtk/auto/container.rs.html#119-126):
```
    fn add<P: IsA<Widget>>(&self, widget: &P) {
        unsafe {
            gtk_sys::gtk_container_add(
                self.as_ref().to_glib_none().0,
                widget.as_ref().to_glib_none().0,
            );
        }
    }
```
It's not using any of the **Rust** methods for widgets - it just converts it to a pointer which gets passed to GTK. I think `IsA` is good for constraining arguments which ensures type safety when you send things to GTK, but doesn't really give you any access to the wrapper methods gtk-rs provides.

Reply:
I see. But if you look for example at who implements `ContainerExt`, then it is says `impl<O: IsA<Container>> ContainerExt for O`. Thus, if you take `IsA<Dialog>` as argument, you **should** get `DialogExt`, `GtkWindowExt`, `ContainerExt`, and a dozen other traits for free. Maybe some imports are missing?

Reply:
I was importing the GTK prelude, so this is not the problem. And Rust usually tells you if the problem is a missing import.

> you **should** get `DialogExt`, `GtkWindowExt`, `ContainerExt`, and a dozen other traits for free

This is not a benefit. I'm not using anything from `ContainerExt` or `BinExt`. Postel's law says "be liberal in what you accept" - every trait I get for free is a demand on the function's users, and if I can avoid these demands I should.